src/main/java/org/openstreetmap/atlas/generator/AtlasGeneratorHelper.java

```diff
@@ -106,115 +95,9 @@ public static ConfiguredTaggableFilter getTaggableFilterFrom(
-                            throw e;
-                        }
-                    }
-
-                    // Before making the move, check again if file is there or not
-                    if (!fileFromTemporaryDirectory.exists())
-                    {
-                        try
-                        {
-                            Files.move(Paths.get(temporaryLocalFile.getPath()),
-                                    Paths.get(fileFromTemporaryDirectory.getPath()),
-                                    StandardCopyOption.ATOMIC_MOVE);
-                        }
-                        catch (final FileAlreadyExistsException e)
-                        {
-                            logger.warn("Failed to rename file, but file exists already.", e);
-                        }
-                        catch (final Exception e)
-                        {
-                            logger.warn("Failed to rename file on local disk.", e);
-                        }
-                    }
-                }
-                catch (final Exception e)
-                {
-                    logger.warn("Failed to cache file on local disk.", e);
-                }
-            }
-
-            // If we were able to find the file on local disk, then load from there
-            if (fileFromTemporaryDirectory.exists())
-            {
-                logger.debug("Atlas exists - Cache Hit: {}",
-                        fileFromTemporaryDirectory.getAbsolutePath());
-                return loadAtlas(fileFromTemporaryDirectory);
-            }
-            else
-            {
-                logger.debug("Falling back to Atlas file hosted on {} for shard {}.",
-                        atlasDirectory, shard.getName());
-                final String path = SparkFileHelper.combine(atlasDirectory,
-                        String.format("%s%s", getAtlasName(country, shard), ATLAS_EXTENSION));
-                final Resource fileFromNetwork = FileSystemHelper.resource(path, sparkContext);
-                try
-                {
-                    return loadAtlas(fileFromNetwork);
-                }
-                catch (final Exception e)
-                {
-                    final Optional<FileNotFoundException> fileNotFound = ExceptionSearch
-                            .find(FileNotFoundException.class).within(e);
-                    if (fileNotFound.isPresent())
-                    {
-                        // It's possible there is no Atlas file for a given shard
-                        logger.debug("No Atlas file found at {}", path);
-                        return Optional.empty();
-                    }
-                    else
-                    {
-                        // There is something else going on, re-throw and continue
-                        throw e;
-                    }
-                }
-            }
+            final HadoopAtlasFileCache atlasCache = new HadoopAtlasFileCache(atlasDirectory,
```

Comment:
Is it correct understanding that each time the fetcher function is invoked, it will create a new `HadoopAtlasFileCache` instance or because we re-use the fetcher for all slicing operations, we are referencing the same underlying cache each time? I guess what I'm asking is, is it possible to persist the cache outside of the Function itself or does it not matter.

Reply:
So in @adahn6 's usage here, it will construct a new cache object (and new caching strategy object) each time. However, since we are always referencing the same underlying filesystem (in the case of a `SystemTemporaryFileCachingStrategy`), and because URIs are mapped to their MD5 sum, the actual cache contents persist across instances of any given `HadoopAtlasFileCache`. The only downside to recycling the object is that you lose the map of precomputed MD5 sums each recycle. So there will be a small performance penalty associated with recomputing the MD5 sum of a URI in order to look it up in the filesystem. However, you will still get cache hits across objects.

Reply:
Makes sense, this may be worth a comment, as it's not obvious by just reading the code. Or perhaps we could pull it out of the function itself, so it's more obvious, your call @adahn6!

Reply:
I am partial to pulling it out into a `static final` class variable, like how we have the `ATLAS_LOADER` class variable for the `AtlasResourceLoader`. `HadoopAtlasFileCache` is threadsafe so it should not be a problem.

Reply:
didn't realize it was thread safe! I instantiated a new object since that seemed safest, but I will pull it out into a `static final` class variable :)

Reply:
wait, I remember why I didn't do that now, several of the function parameters are required for the HadoopAtlasFileCache constrctor (spark context and atlas directory). because of that, there's not a good way to make them `static final`. I could, however, make one cache object and pass it in to the function?

Reply:
That sounds reasonable.

Reply:
Ah shoot that's right. Originally, I designed the cache classes (`ConcurrentResourceCache` and `HadoopAtlasFileCache`) to allow you to change the parameters with some `set` methods. But then this compromised thread safety because there would be a race between the setters and the caches' `get` methods (since a `get` could preempt one of the `set` calls before it can properly complete).

One way to get around this would be to have the constructor take a lock object, which the user would have to pass in themselves. You could then lock the `set` and `get` methods with that common lock. I decided against this since it makes the class more complicated. However, as these classes are integrated into our codebase, maybe it makes more sense to change the design?

Reply:
changing the directory sounds like it would make things really complicated for the MD5 map-- without updating it too, the mapping is probably invalidated, and updating that map wouldn't necessarily be fast or easy. IMO, I think the current implementation is good for now. As use cases expand, we can examine more complex projects with multiple working directories to see what's best. 
Off the top of my head, I'd consider a `CacheManager` that takes requests for caches and gives back an instantiated object; abstracted in that class would be logic around "if a cache exists, return it else make a new one for that directory".

Reply:
wait nm, I was misreading some code and thought it was just hashing based off filename

Reply:
It's based off the full URI. You're not the first person to mix that up :)

Reply:
could we just switch the cache to use URIs then? seems like a pretty minor change overall, but allows for the reuse of one cache object everywhere

Reply:
if we wanted to keep filename to URI conversion logic in the cache itself, could make it a static method that takes in filename, base dir, and spark config and returns the expected URI